### ISSUE TITLE
Remove file path and line in Bline

### ISSUE THREAD

**Issue by darzok0914** [ibhagwan/fzf-lua#228]
how should I configure BLines if I don't want to have the file name and line number displayed ? I'm asking because if I have a file named xx and I want to search for xx within that file then it doesn't work as basically every line of that file matches since the file name are part of the result.

**Comment by ibhagwan** [comment]
You need to supply fzf's `--nth` to exclude fuzzy search or `--with-nth` to hide alltogether.

[From the Wiki: Disable or hide filename fuzzy search](https://github.com/ibhagwan/fzf-lua/wiki#nth)

Since `blines` already uses the `]` delimiter to exclude the buffer ID from the search you need to add the `:` to the delimiter and hide anything before field 3:
```lua
:lua require'fzf-lua'.blines({ fzf_cli_args = '--delimiter="[\\]:]" --with-nth 3..' })
```

**Comment by darzok0914** [comment]
Hi, 

thanks for your response, I tried it and at first it seemed like it did the trick, however it is not working as expected. 
when I do this command on a file called main.cpp and I look for a string main() this is what I have :


![image](https://user-images.githubusercontent.com/72544332/144225822-85070d78-3742-47c3-ae76-0114920e5042.png)

you can see that it finds the result line 64 but not the one line 57 as it should, however the normal BLines would find it, I was thinking of passing some options to rg directly from the BLines command but I can seem to make it working

I tried something like this:
:lua require'fzf-lua'.blines({rg_opts= '--column --no-filename --line-number --no-heading --color=always --smart-case'})

but it didn't work

**Comment by ibhagwan** [comment]
Turns out that when specifying `--with-nth` you also need to redifine the fuzzy search with `--nth`, otherwise it seems that it will only fuzzy search the first field between the `:`, the below worked well for me (also removed the `]` separator as it wasn't needed:
```lua
-- `--delimiter="[:]` - separate fields by colon
-- `--with-nth=2..`   - display data only from after the first colon
-- `--nth=1..`        - fuzzy search all fields regardless of colon
:lua require'fzf-lua'.blines({ fzf_cli_args = '--delimiter="[:]" --with-nth=2.. --nth=1..' })
```

As for:

>I tried something like this:
> :lua require'fzf-lua'.blines({rg_opts= '--column --no-filename --line-number --no-heading --color=always --smart-case'})

This was not going to work as `blines` (and `lines`) **does not use `rg`** but actual buffer lines retrieved using the `vim.api.nvim_buf_get_lines` API, for this puprose this is much better than `rg` as it doesn't require the buffer to be saved on disk meaning you can find lines on "dirty" buffers and buffers that weren't even saved on disk.

FYI, only the `grep` and `live_grep` providers user an underlying `rg` command (which can be customized), if you wish to use `rg` (or `grep`) on a single file you can still do it with:
```lua
:lua require'fzf-lua'.grep_curbuf()
```

**Comment by ibhagwan** [comment]
`--tiebreak=index` option seems to do what you want:
```lua
:lua require'fzf-lua'.blines({ fzf_cli_args = '--delimiter="[:]" --with-nth=2.. --nth=1.. --tiebreak=index' })
```

From fzf's man page:
```help
       --tiebreak=CRI[,..]
              Comma-separated list of sort criteria to apply when the scores
              are tied.

              length  Prefers line with shorter length
              begin   Prefers line with matched substring closer to the
              beginning
              end     Prefers line with matched substring closer to the end
              index   Prefers line that appeared earlier in the input stream

              - Each criterion should appear only once in the list
              - index is only allowed at the end of the list
              - index is implicitly appended to the list when not specified
              - Default is length (or equivalently length,index)
              - If end is found in the list, fzf will scan each line backwards
```

Btw, on the latest commit (a few minutes ago) I added a new "shortcut" command `lgrep_curbuf` (to the already existing `grep_curbuf`) both use `rg` as the underlying command so you can play around with `rg_opts` for the current buffer as you tried to do in an earlier post, I still believe `blines` is superior for what you want but `lgrep_curbuf` ("live grep" the current buffer) gives you the option to perform a regex search which you cannot do in `blines`'s fuzzy search.

An important edit I added to an earlier post:

> If you wish to understand what's the differnce between `grep` and `live_grep` read [What's the difference between grep and live_grep?](https://github.com/ibhagwan/fzf-lua/wiki#grep-vs-live-grep)

**Comment by ibhagwan** [comment]
FYI, latest commit (https://github.com/ibhagwan/fzf-lua/commit/a5e1be796d8905d96028dbf9567dc266308239a1) adds `--tiebreak=index` to `lines` and `blines` by default so you don't have to add it anymore.

**Comment by darzok0914** [comment]
Nice, thanks a lot for the support, btw I tried to delete and reinstall the plugin to see if this option was working by default and it did not work but it might be something I did wrong. either way I mapped to the full cmd : 

nnoremap <leader>L :lua require'fzf-lua'.blines({ fzf_cli_args = '--delimiter="[:]" --with-nth=2.. --nth=1.. --tiebreak=index', show_unlisted=true })<cr>

is there any reason why the show_unlisted=true is not by default ? my understanding is that I can BLines the vim help menu without it but is there any downsides for having it enabled all the time?

**Comment by ibhagwan** [comment]
> Nice, thanks a lot for the support, btw I tried to delete and reinstall the plugin to see if this option was working by default and it did not work but it might be something I did wrong. either way I mapped to the full cmd :
> 
> nnoremap L :lua require'fzf-lua'.blines({ fzf_cli_args = '--delimiter="[:]" --with-nth=2.. --nth=1.. --tiebreak=index', show_unlisted=true })

It should work, you can check the commit version of your plugin in the folder it was cloned into which would depend on the plugin manager you use, for exmaple, on my machine with `packer.nvim` the path is: `.local/share/nvim/site/pack/packer/opt/fzf-lua/lua` but if you're using `vim-plug` it will be a different path.

You don't need to uninstall and reinstall in order to get the latest version, plugin managers usually have a command to get the latest changes, if you're using `vim-plug` you can just run `:PlugUpgrade`.

> is there any reason why the show_unlisted=true is not by default ? my understanding is that I can't BLines the vim help menu without it but is there any downsides for having it enabled all the time?

The buffer listing code is the same across all buffer related commands (`buffers`, `tabs`, `lines`, `blines`) where it wouldn't make sense to show unlisted buffers (it's not only `help`), you're probably right though that `blines` is different as it doesn't matter which buffer you're on it should be visible.

https://github.com/ibhagwan/fzf-lua/commit/253b9413eb116081aa81becda67ed9ccbccbb0a6  - this commit sets the defaults for `blines`:  `{ show_unlisted = true, no_term_buffers = false }` (can be controlled via the `setup()` function).

> side note regarding the `help` buffer, I never understood why it's an unlisted buffer, IMO it sohuld be listed in a similar way that `:Man` buffers are listed.

**Comment by darzok0914** [comment]
- "You don't need to uninstall and reinstall in order to get the latest version, plugin managers usually have a command to get the latest changes, if you're using vim-plug you can just run :PlugUpgrade." I'm using vim-plug indeed but I didn't specify the commit hash of some of the newest plugin I have and I don't want to update them if they are stable and work fine. 

- this commit sets the defaults for blines: { show_unlisted = true, no_term_buffers = false } (can be controlled via the setup() function). Thanks for this quick upgrade, and also the explanation :)

I was wondering if somehow you knew why when we do a BLines in the vim-help buffer the preview window doesn't display any colors?

**Comment by ibhagwan** [comment]
> * "You don't need to uninstall and reinstall in order to get the latest version, plugin managers usually have a command to get the latest changes, if you're using vim-plug you can just run :PlugUpgrade." I'm using vim-plug indeed but I didn't specify the commit hash of some of the newest plugins I have and I don't want to update them if they are stable and work fine.

Understood regarding the auto-updates, althought with the current speed of developent in neovim it would be worthwhile checking out new features and also compatibility issues if you've upgraded to 0.6 or using the nightly.

> I was wondering if somehow you knew why when we do a BLines in the vim-help buffer the preview window doesn't display any colors?

That's because help buffers are simple `txt` files and the code was setting `filetype=txt`.

https://github.com/ibhagwan/fzf-lua/commit/1d6ea8d38100b9b4b4f1e49626697c9c41a6759c - this commit fixes that as it copies the filetype from the exisiting buffer.

In addition, hiding the filename makes more sense than not for a single buffer operation so I made that the default in the above commit (so you no longer have to specify these options with `fzf_cli_args`). If you wish to change the defaults they now appear in the [README cusomization section](https://github.com/ibhagwan/fzf-lua#customization) under `blines.fzf_opts` and `lines.fzf_opts` respectively.

Basically, all your suggestions/customizations were implemented by default, it should work the way you wanted it without any option added, ty for your dilligent testing :-)

**Comment by darzok0914** [comment]
"Basically, all your suggestions/customizations were implemented by default, it should work the way you wanted it without any option added, ty for your dilligent testing :-)"

yes I've noticed, you're really fast as well at responding it just blows my mind, thanks so much it's just incredible. 

I've updated to your latest version and indeed it fixes the colours for the vim-help menu. however it is not working on my cpp files
![image](https://user-images.githubusercontent.com/72544332/144496293-606c2ae3-b07e-4dc1-afa4-2f2e844b6e66.png)

but it works for my bash, lua and python files. 

I'm still running nvim 0.5.1 and besides your plugin the only thing I've updated is tree sitter. do you have any idea what it could be?

**Comment by darzok0914** [comment]
I just updated to nvim 0.6.0 just to be sure and I have a similar issue

**Comment by ibhagwan** [comment]
> I just updated to nvim 0.6.0 just to be sure and I have a similar issue

My fix should have worked regardless of the filetype as it simply copies the filetype from the source (filetype sets the highlights), I'm not sure why this messed up the filetype, we can try to fix it though or I apply the filetype workaround to 'txt' files only.

I have some issue with treesitter I can't even get highlights for `cpp` in neovim at all (not just preview), can you run `:set ft=?` inside your `cpp` file and let me know what's the file type for it?

**Comment by darzok0914** [comment]
are you sure about this command :set ft=? because when I try this I have:

![image](https://user-images.githubusercontent.com/72544332/144499448-68871990-8347-4f5b-812a-258098f8555d.png)

**Comment by ibhagwan** [comment]
whoops I meant `:set ft?`, just tryin to figure out what the filetype is, it should reply either `c` or `cpp`.

**Comment by darzok0914** [comment]
oki, I run :verbose set ft ? and I got 

![image](https://user-images.githubusercontent.com/72544332/144499841-9de7c541-3d22-4a8f-8a38-6bca59492bbc.png)

**Comment by ibhagwan** [comment]
I just tested it myself too and I get the highlights working properly for `cpp`, before the change the filetype would be auto-detected from the filname+content (which should end up as you've described `cpp`) but that also means that getting the filetype from buffer should also set it to `cpp` ending with the same syntax highlihgted result, not sure why this happens.

On top of that it seems to work on my end, definitely confusing.

![screenshot-1638477879](https://user-images.githubusercontent.com/59988195/144500458-a05a534d-c26a-4117-a87b-e9dc66c16aa6.png)

**Comment by ibhagwan** [comment]
Just used another machine I have with older version of treesitter where I don't have the `cpp` treesitter errors, and it seems to work (note the distinct colors for `main` and `print`).

I'd rather avoid creating an ugly workaround unless I undertand what the real issue is, theoretically copying the `filetype` from the source buffer should always work better than any other method - are you having any issues with `cpp` highlighting in general or just in the preview?

![screenshot-1638478432](https://user-images.githubusercontent.com/59988195/144501506-af9753ba-d7f2-4071-9c46-b92c95820fb5.png)

**Comment by ibhagwan** [comment]
I was just able to fix my treesitter by removing the entire `~/.local/share/nvim` folder and reinstalling all my plugins (along with treesitter) and everything seems to work fine for me with 3 different neovim versions 0.5.1, 0.6 and the 0.7-nightly (perhaps something to do with the latest upgrade to 0.6).

> Note that this will delete your shared data (AKA ShaDa), if you wish the keep your command/register history, etc make sure to backup and restore `~/.local/share/nvim/shada`

Previews for `cpp` seem to work as expected (note that extra colurful treesitter highlights):

![screenshot-1638479307](https://user-images.githubusercontent.com/59988195/144503501-78746450-4c81-4245-9ed3-5d545907b7af.png)

**Comment by darzok0914** [comment]
oki, so I checkout this version: 253b9413eb116081aa81becda67ed9ccbccbb0a6 of your plugin and it works well
and the 'bug' was really introduced in this commit 1d6ea8d38100b9b4b4f1e49626697c9c41a6759c

**Comment by ibhagwan** [comment]
> oki, so I checkout this version: [253b941](https://github.com/ibhagwan/fzf-lua/commit/253b9413eb116081aa81becda67ed9ccbccbb0a6) of your plugin and it works well and the 'bug' was really introduced in this commit [1d6ea8d](https://github.com/ibhagwan/fzf-lua/commit/1d6ea8d38100b9b4b4f1e49626697c9c41a6759c)

Yes, that's the help file workaround, not sure why setting the filetype manuall to `cpp` doesn't work for you.

https://github.com/ibhagwan/fzf-lua/commit/04912f6eff4dc2b9a40f1774cc4ac8a5bdec9405 - latest commit limits the filetype workaround to `help` files only, this should solve everything for you.

Not totally crazy about this as we didn't figure out why this happens but given there are very limited occurances when file extensions do not match their filetypes I guess I can live with that :-)

Try it out, I'm pretty sure this will solve everything for you.

**Comment by darzok0914** [comment]
I deleted my ~/.local/share/nvim and reinstalled everything, it did not fix the issue. 

I can confirm that the latest commit fix the problem :)

**Comment by ibhagwan** [comment]
> I deleted my ~/.local/share/nvim and reinstalled everything, it did not fix the issue.
> 
> I can confirm that the latest commit fix the problem :)

Alright then, I guess we can close this issue and move to #229 :-)

**Comment by ibhagwan** [comment]
Btw, if you wanna test why it didn't work, we can do what the previewer does manually.

Open a new file using `nvim` (with no filename), and input the following text inside:
```cpp
int main() {
    print("test")
}
```
The text should have no syntax highlighting, then run `:set ft=cpp` and the text should become highlgihted, if you feel like it and get a chance let me know if this works?

**Comment by darzok0914** [comment]
yep, it does exactly what you said, no syntax highlighting before setting the file type. 

I'm closing the issue, tbh I feel like we have gone through a lot on that issue :) thx for your help, you've been amazing
